### PR TITLE
Improve performance of string.replace

### PR DESCRIPTION
**Pull request by Quarz0**
This PR improves the performance, in terms of time, of `replace` method by replacing regex by linear scanning which is much faster.
Using the benchmark [here](https://github.com/Quarz0/starlark/blob/benchmark_runner/benchmark_suite/data/string_replace.star), the results obtained on my local machine were as follows:
```
=== data/string_replace.star ===
                   Time (MS)  Memory (KB)
java               10777.098    89340.000    # Before
java                1728.092   184904.000    # After
go                    10.259     4496.000
rust                6615.044     4500.000
python3               50.873     5640.000
```

~~This also fixes the case where `maxsplits` is a negative value. According to the [spec](https://github.com/bazelbuild/starlark/blob/master/spec.md#stringreplace), a negative value for the number of splits should be ignored and this is also what python and the Go implementation do.~~
```python
# Before
"abc".replace("a", "z", -1)    # "abc"
# After
"abc".replace("a", "z", -1)    # "zbc"
```

### CODE REVIEW

**Comment by laurentlb** [comment]
cc @alandonovan, @meisterT

**Comment by laurentlb** [comment]
Does your time column include the startup time? If so, it's hard to make language comparisons. You could check the time to run a "Hello World" and subtract that.

**Comment by meisterT** [comment]
> Does your time column include the startup time? If so, it's hard to make language comparisons. You could check the time to run a "Hello World" and subtract that.

Yes, it does include the startup time, see also my comments on https://github.com/bazelbuild/starlark/pull/75

**Comment by Quarz0** [comment]
> Does your time column include the startup time? If so, it's hard to make language comparisons. You could check the time to run a "Hello World" and subtract that.

I guess subtracting the startup time is a good idea, but I thought in this case, just comparing the before and after performance of Java alone would be a good indicator of performance improvement, regardless of comparing Java with other implementations.

**Comment by laurentlb** [comment]
Yes, the first two lines show a very nice performance improvement, thanks!
@meisterT will run additional benchmarks (in our codebase) to measure the impact.

Can you separate the performance improvement from the semantic change (negative number of splits)? We might want an incompatible flag for that, even if a breakage is unlikely.

**Comment by Quarz0** [comment]
> Can you separate the performance improvement from the semantic change (negative number of splits)? We might want an incompatible flag for that, even if a breakage is unlikely.

Done. I also fixed the case when `oldString` is the empty string which was incorrectly handled in the previous commit. There shouldn't be any changes in semantics now.

File: src/main/java/com/google/devtools/build/lib/syntax/StringModule.java [hunk 1]
```diff
@@ -281,25 +282,29 @@ public String strip(String self, Object charsOrNone) {
...
+      if (maxSplit > 0) {
+        sb.append(newString);
       }
-      m.appendTail(sb);
-    } catch (IllegalStateException e) {
-      throw new EvalException(loc, e.getMessage() + " in call to replace");
+      return sb.toString();
     }
-    return sb.toString();
+    return StringUtils.replace(self, oldString, newString, maxSplit);
```
Comment (alandonovan) [inline src/main/java/com/google/devtools/build/lib/syntax/StringModule.java]: The replace function needed here is pretty easy to write; I don't think it warrants another dependency:

```
        int start = 0;
        for (int i = 0; i < maxSplit; i++) {
            int end = self.indexOf(old, start);
            if (end < 0) break;
            sb.append(self, start, end).append(newString);
            start = end + old.length();
        }
        sb.append(self, start, self.length());
```

Comment (Quarz0) [reply src/main/java/com/google/devtools/build/lib/syntax/StringModule.java]: I was actually planning to use `StringUtils` in other methods too like `count` and `split` to improve their performance as well.

Comment (alandonovan) [reply src/main/java/com/google/devtools/build/lib/syntax/StringModule.java]: To be clear: the performance improvement in this change comes not from anything clever in the StringUtils package, but from avoiding the regexp package, which has substantial overheads. By all means, avoid the regexp package in count and split too, but I don't think there's any need for an additional dependency to do that. Both functions can be implemented as a simple loop around indexOf. For example:
```
    List<String> split(String str, String sep, int max) {
        List<String> res = new ArrayList<>();
        int n = 0, i = 0;
        for (;;) {
            int j = str.indexOf(sep, i);
            if (j < 0 || ++n == max) {
                res.add(str.substring(i));
                break;
            }
            res.add(str.substring(i, j));
            i = j + sep.length();
        }
        return res;
    }
```

File: src/test/starlark/testdata/string_misc.sky [hunk 2]
```diff
@@ -41,6 +41,18 @@ assert_eq('b$()n$()n$()'.replace('$()', '$($())'), "b$($())n$($())n$($())")
 assert_eq('b\\n\\n\\'.replace('\\', '$()'), "b$()n$()n$()")
 
 assert_eq('banana'.replace('a', 'e', 2), "benena")
+assert_eq('banana'.replace('a', 'e', 0), "banana")
+assert_eq('banana'.replace('a', 'e', -1), "banana")
```
Comment (alandonovan) [inline src/test/starlark/testdata/string_misc.sky]: Is this right? I would expect benene.

File: src/test/starlark/testdata/string_misc.sky [hunk 3]
```diff
@@ -41,6 +41,18 @@ assert_eq('b$()n$()n$()'.replace('$()', '$($())'), "b$($())n$($())n$($())")
...
+assert_eq('banana'.replace('a', 'e', -10), "banana")
```
Comment (alandonovan) [inline src/test/starlark/testdata/string_misc.sky]: same here

File: src/test/starlark/testdata/string_misc.sky [hunk 4]
```diff
@@ -41,6 +41,18 @@ assert_eq('b$()n$()n$()'.replace('$()', '$($())'), "b$($())n$($())n$($())")
...
+
+assert_eq('banana'.replace('', '-'), "-b-a-n-a-n-a-")
+assert_eq('banana'.replace('', '-', -2), "banana")
```
Comment (alandonovan) [inline src/test/starlark/testdata/string_misc.sky]: same here

File: src/main/java/com/google/devtools/build/lib/syntax/StringModule.java [hunk 5]
```diff
@@ -281,24 +281,32 @@ public String strip(String self, Object charsOrNone) {
...
-      for (int i = 0; i < maxSplit && m.find(); i++) {
-        m.appendReplacement(sb, Matcher.quoteReplacement(newString));
+    int maxSplit = Integer.MAX_VALUE;
+    if (maxSplitO != Runtime.NONE) {
+      maxSplit = Math.max(0, (Integer) maxSplitO);
+    }
+    StringBuilder sb = new StringBuilder();
+    int start = 0;
+    for (int i = 0; i < maxSplit; i++) {
+      if (oldString.length() == 0) {
```
Comment (alandonovan) [inline src/main/java/com/google/devtools/build/lib/syntax/StringModule.java]: oldString.isEmpty()

File: src/test/starlark/testdata/string_misc.sky (hunk 2 above)
Comment (Quarz0) [reply src/test/starlark/testdata/string_misc.sky]: Yes, that's a bug in the current implementation, but as Laurent mentioned above that we'd better not include semantic changes in this PR.

File: src/main/java/com/google/devtools/build/lib/syntax/StringModule.java (hunk 5 above)
Comment (Quarz0) [reply src/main/java/com/google/devtools/build/lib/syntax/StringModule.java]: Done

File: src/test/starlark/testdata/string_misc.sky (hunk 2 above)
Comment (alandonovan) [reply src/test/starlark/testdata/string_misc.sky]: Ah, so he did.

In general if you identify a bug but don't intend to fix it in the same PR, it's good practice to at least identify the problem with comments (both at the logic and at the faulty assertions). But let's not hold up this PR any further. LGTM

**Pull request closed without merging**